File: source/com/barrybecker4/simulation/habitat/creatures/Populations.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.habitat.creatures;

import com.barrybecker4.common.math.function.CountFunction;
import com.barrybecker4.common.math.function.Function;
import com.barrybecker4.simulation.habitat.model.Cell;
import com.barrybecker4.simulation.habitat.model.HabitatGrid;
import com.barrybecker4.ui.renderers.MultipleFunctionRenderer;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * Create populations for all our creatures.
 *
 * @author Barry Becker
 */
public abstract class Populations extends ArrayList<Population> {

    private long dayCount = 0;

    /** associate population with function*/
    private Map<Population, CountFunction> functionMap;

    private HabitatGrid grid;

    /**
     * Constructor
     */
    public Populations() {

        initialize();
    }

    public void initialize() {
        functionMap = new HashMap<Population, CountFunction>();
        grid = new HabitatGrid(20, 15);

        this.clear();
        addPopulations();

        updateGridCellCounts();
    }

    protected abstract void addPopulations();

    public void nextDay() {
        for (Population pop : this) {
            pop.nextDay(grid);
        }
        // remove any creatures that might have died by starvation or being eaten.
        for (Population pop : this) {
            pop.removeDead(grid);
        }
        updateFunctions(dayCount);
        dayCount++;
    }

    public MultipleFunctionRenderer createFunctionRenderer() {

        List<Function> functions = new ArrayList<Function>();
        List<Color> lineColors = new LinkedList<Color>();

        for (Population pop : this) {
            CountFunction func = new CountFunction(pop.getSize());
            functions.add(func);
            lineColors.add(pop.getType().getColor());

            functionMap.put(pop, func);
        }

        return new MultipleFunctionRenderer(functions, lineColors);
    }

    private void updateFunctions(long iteration) {

        for (Population pop : this) {
            functionMap.get(pop).addValue(iteration, pop.getSize());
        }
    }

    private void updateGridCellCounts() {
        for (Population pop : this) {
            for (Creature c : pop.getCreatures()) {
                Cell cell = grid.getCellForPosition(c.getLocation());
                cell.addCreature(c);
            }
        }
    }

}

File: scala-source/com/barrybecker4/simulation/habitat/creatures/Populations.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.habitat.creatures

import java.awt.Color
import com.barrybecker4.math.function.CountFunction
import com.barrybecker4.math.function.Function
import com.barrybecker4.simulation.habitat.model.HabitatGrid
import com.barrybecker4.ui.renderers.MultipleFunctionRenderer
import scala.collection.mutable.ArrayBuffer


/**
  * Create populations for all our creatures.
  * @author Barry Becker
  */
abstract class Populations extends ArrayBuffer[Population] {

  private var functionMap = Map[Population, CountFunction]()
  private var dayCount: Int = 0
  /** associate population with function */
  private var grid: HabitatGrid = _
  initialize()

  private def initialize(): Unit = {
    grid = new HabitatGrid(20, 15)
    this.clear()
    addPopulations()
    updateGridCellCounts()
  }

  def reset(): Unit = {
    grid = new HabitatGrid(20, 15)
    for (pop <- this) pop.reset()
    updateGridCellCounts()
  }

  protected def addPopulations(): Unit

  def nextDay(): Unit = {
    for (population <- this) population.nextDay(grid)

    // remove any creatures that might have died by starvation or by being eaten.
    for (pop <- this) pop.removeDead()

    updateFunctions(dayCount)
    
    dayCount += 1
  }

  def createFunctionRenderer: MultipleFunctionRenderer = {
    var functions = List[Function]()
    var lineColors: List[Color] = List[Color]()

    for (pop <- this) {
      val func = new CountFunction(pop.getSize)
      functions +:= func
      lineColors +:= pop.creatureType.color
      functionMap += pop -> func
    }
    val funcRenderer = new MultipleFunctionRenderer(functions, Some(lineColors))
    funcRenderer.setRightNormalizePercent(40)
    funcRenderer
  }

  private def updateFunctions(iteration: Long): Unit = {
    for (pop <- this) {
      val func = functionMap(pop)
      func.addValue(iteration.toDouble, pop.getSize.toDouble)
    }
  }

  private def updateGridCellCounts(): Unit = {
    for (pop <- this) {
      for (c <- pop.creatures) {
        val cell = grid.getCellForPosition(c.getLocation)
        cell.addCreature(c)
      }
    }
  }
}
